crates/bevy_platform_support/src/staging.rs

```diff
@@ -0,0 +1,460 @@
+        Stager {
+            cold: &mut self.cold,
+            staged: self.staged.get_mut().unwrap(),
+        }
+    }
+
+    /// Constructs a [`StagerLocked`], locking internally.
+    ///
+    /// # Deadlocks
+    ///
+    /// This deadlocks if there are any other lock guards on this thread for this value.
+    #[inline]
+    pub fn stage_lock(&self) -> StagerLocked<'_, T, &T::Cold> {
+        StagerLocked {
+            cold: &self.cold,
+            staged: self.staged.write().unwrap(),
+        }
+    }
+
+    /// Constructs a [`Stager`] that will stage changes.
+    #[inline]
+    pub fn read(&mut self) -> StagedRef<'_, T> {
+        StagedRef {
+            cold: &self.cold,
+            staged: self.staged.get_mut().unwrap(),
+        }
+    }
+
+    /// Constructs a [`StagedRefLocked`], locking internally.
+    ///
+    /// # Deadlocks
+    ///
+    /// This deadlocks if there are any write lock guards on this thread for this value.
+    #[inline]
+    pub fn read_lock(&self) -> StagedRefLocked<'_, T, &T::Cold> {
+        StagedRefLocked {
+            cold: &self.cold,
+            staged: self.staged.read().unwrap(),
+        }
+    }
+
+    /// Runs different logic depending on if additional changes are already staged.
+    /// This can be faster than greedily applying staged changes if there are already staged changes.
+    pub fn maybe_stage<C, S>(
+        &mut self,
+        for_full: impl FnOnce(&mut T::Cold) -> C,
+        for_staged: impl FnOnce(&mut Stager<T>) -> S,
+    ) -> MaybeStaged<C, S> {
+        let staged = self.staged.get_mut().unwrap();
+        let cold = &mut self.cold;
+        if staged.any_staged() {
+            MaybeStaged::Staged(for_staged(&mut Stager { cold, staged }))
+        } else {
+            MaybeStaged::Cold(for_full(cold))
+        }
+    }
+
+    /// Easily run a stager function to stage changes.
+    ///
+    /// # Deadlocks
+    ///
+    /// This deadlocks if there are any other lock guards on this thread for this value.
+    #[inline]
+    pub fn stage_scope_locked<R>(&self, f: impl FnOnce(&mut Stager<T>) -> R) -> R {
+        f(&mut self.stage_lock().as_stager())
+    }
+
+    /// Easily run a [`StagedRef`] function.
+    ///
+    /// # Deadlocks
+    ///
+    /// This deadlocks if there are any write lock guards on this thread for this value.
+    #[inline]
+    pub fn read_scope_locked<R>(&self, f: impl FnOnce(&StagedRef<T>) -> R) -> R {
+        f(&self.read_lock().as_staged_ref())
+    }
+}
+
+impl<T: StagedChanges> AtomicStageOnWrite<T> {
+    /// Constructs a new [`AtomicStageOnWrite`] with the given value and no staged changes.
+    pub fn new(value: T::Cold) -> Self {
+        Self(Arc::new(AtomicStageOnWriteInner {
+            cold: RwLock::new(value),
+            staged: RwLock::default(),
+        }))
+    }
+
+    /// Gets the inner cold data if it is safe. If [`any_staged`](Self::any_staged) is known to be false, this can be safely unwrapped.
```

Comment:
```suggestion
    /// Gets the inner cold data if it is safe.
    /// If [`any_staged`](Self::any_staged) is known to be false, this can be safely unwrapped.
```
See above regarding use of the term "safe".